RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc

```diff
@@ -0,0 +1,413 @@
+  }
+
+  for (auto ittkmu = muColl->begin(); ittkmu != muColl->end(); ittkmu++) {
+    // L1 tracker track
+    auto it = ittkmu->trkPtr();
+
+    // propagate to GMT
+    auto p3 = it->momentum();
+    float tk_pt = p3.perp();
+    float tk_p = p3.mag();
+    float tk_eta = p3.eta();
+    float tk_aeta = std::abs(tk_eta);
+    float tk_phi = p3.phi();
+    float tk_q = it->rInv() > 0 ? 1. : -1.;
+    float tk_z = it->POCA().z();
+
+    if (tk_p < theMinPL1Tk)
+      continue;
+    if (tk_aeta < 1.1 && tk_pt < theMinPtL1TkBarrel)
+      continue;
+
+    float dzCorrPhi = 1.;
+    float deta = 0;
+    float etaProp = tk_aeta;
+
+    if (tk_aeta < 1.1) {
+      etaProp = 1.1;
+      deta = tk_z / 550. / cosh(tk_aeta);
+    } else {
+      float delta = tk_z / 850.;  //roughly scales as distance to 2nd station
+      if (tk_eta > 0)
+        delta *= -1;
+      dzCorrPhi = 1. + delta;
+
+      float zOzs = tk_z / 850.;
+      if (tk_eta > 0)
+        deta = zOzs / (1. - zOzs);
+      else
+        deta = zOzs / (1. + zOzs);
+      deta = deta * tanh(tk_eta);
+    }
+    float resPhi = tk_phi - 1.464 * tk_q * cosh(1.7) / cosh(etaProp) / tk_pt * dzCorrPhi - M_PI / 144.;
+    resPhi = reco::reduceRange(resPhi);
+
+    float pt = tk_pt;  //not corrected for eloss
+    float eta = tk_eta + deta;
+    float theta = 2 * atan(exp(-eta));
+    float phi = resPhi;
+    int charge = it->rInv() > 0 ? 1 : -1;
+
+    bool barrel = tk_aeta < 1.1 ? true : false;
+
+    if (pt < theL1MinPt || fabs(eta) > theL1MaxEta)
+      continue;
+
+    LogDebug(metname) << "New L2 Muon Seed";
+    LogDebug(metname) << "Pt = " << pt << " GeV/c";
+    LogDebug(metname) << "eta = " << eta;
+    LogDebug(metname) << "theta = " << theta << " rad";
+    LogDebug(metname) << "phi = " << phi << " rad";
+    LogDebug(metname) << "charge = " << charge;
+    LogDebug(metname) << "In Barrel? = " << barrel;
+
+    // Update the services
+    theService->update(iSetup);
+
+    const DetLayer *detLayer = nullptr;
+    float radius = 0.;
+
+    CLHEP::Hep3Vector vec(0., 1., 0.);
+    vec.setTheta(theta);
+    vec.setPhi(phi);
+
+    DetId theid;
+    // Get the det layer on which the state should be put
+    if (barrel) {
+      LogDebug(metname) << "The seed is in the barrel";
+
+      // MB2
+      theid = DTChamberId(0, 2, 0);
+      detLayer = theService->detLayerGeometry()->idToLayer(theid);
+      LogDebug(metname) << "L2 Layer: " << debug.dumpLayer(detLayer);
+
+      const BoundSurface *sur = &(detLayer->surface());
+      const BoundCylinder *bc = dynamic_cast<const BoundCylinder *>(sur);
+
+      radius = fabs(bc->radius() / sin(theta));
+
+      LogDebug(metname) << "radius " << radius;
+
+      if (pt < theMinPtBarrel)
+        pt = theMinPtBarrel;
+    } else {
+      LogDebug(metname) << "The seed is in the endcap";
+
+      // ME2
+      theid = theta < Geom::pi() / 2. ? CSCDetId(1, 2, 0, 0, 0) : CSCDetId(2, 2, 0, 0, 0);
```

Comment:
it might be better to use `M_PI_2` instead of `Geom::pi() / 2.`, but I don't know if there are any guidelines about this.